OpenTrons/opentrons-api
# feat(api): Pause run after recoverable errors

# Overview

This closes most of EXEC-301 and closes EXEC-320.

# Changelog

## What works now (or is intended to work, anyway)

* When running a JSON protocol on a Flex, if a `pickUpTip` command fails because of a missing tip, the run pauses now, instead of homing and showing an error message.
* You can then either:
    * Cancel the run in the usual way, by issuing a `stop` action via `POST /runs/{id}/actions`.
    * Resume the run from the next command, by issuing a `resume-from-recovery` action via `POST /runs/{id}/actions`.

## What doesn't work

* After a `pickUpTip` command failure, the robot seems to simultaneously think it does not have a tip attached (a subsequent `aspirate` command will fail with a "tip not attached" error) and that it does have a tip attached (a subsequent `moveTo` will act like there's a tip). I haven't had a chance to look into this yet. So the practical usefulness of this is pretty limited so far: you can only meaningfully recover if the protocol doesn't have any aspirates or dispenses.
* After recovering from an error and proceeding through the rest of the run, the run will end as `failed` despite the recovery. I see what's causing this, I just haven't gotten to it yet.
* You can't issue any fixup commands yet.
* You can't recover from errors in Python protocols yet. EXEC-339 is the next step to accomplish that.
* This isn't remotely supported by the Opentrons App or Flex on-device display. You'll see weird things like buttons without labels.

# Test Plan

## Setup

Enable the hidden error recovery feature flag by manually issuing a `POST /settings` request with `{"id": "enableErrorRecoveryExperiments", "value": true}`.

## Test

1. Upload this test protocol to a Flex: [Recovery test.json.zip](https://github.com/Opentrons/opentrons/files/14654980/Recovery.test.json.zip). This comes from Protocol Designer, with `aspirate`s and `dispense`s hand-replaced by `moveToWell`s in order to work around the tip-attached confusion that I've described above.
2. Fill random wells in the leftmost column of the tip rack with tips.
3. Start the run.
4. When the run reaches a missing tip, it should pause. Resume it by issuing a `{"data": {"actionType": "resume-from-recovery"}}` action to `POST /runs/{id}/actions`, or cancel it by issuing a `{"data": {"actionType": "stop"}}` action.


## Cleanup

1. Delete any runs that you just created, to avoid database incompatibilities when you push a different branch. You can do this with the Opentrons App or with manually-issued `DELETE /runs/{id}` requests.
5. `POST /settings` with `{"id": "enableErrorRecoveryExperiments", "value": false}`.

# Review requests

None in particular.

The easiest way to review this might be to filter by commit. I've organized this into refactor commits vs. feature commits.

# Risk assessment

Medium. This touches core internal Protocol Engine logic. Although it's all technically well covered by unit tests, those tests are mostly too isolated and low-level to catch actual bugs.
